compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -51,41 +51,108 @@ static bool hasIntrinsic(IREE::GPU::TargetAttr target,
 }
 
 static std::optional<IREE::GPU::DataTiledMMAAttr>
-chooseDataTiledMMAAttr(TypeRange elementTypes, IREE::GPU::TargetAttr target) {
-  assert(elementTypes.size() == 3);
+chooseDataTiledMMAAttr(TypeRange eTypes, IREE::GPU::TargetAttr target,
+                       IREE::Encoding::EncodingAttr encoding) {
   using namespace IREE::GPU;
   MLIRContext *ctx = target.getContext();
-  Type lhs = elementTypes[0];
-  Type rhs = elementTypes[1];
-  Type out = elementTypes[2];
-  auto match = [=](MMAIntrinsic intrinsic, int unrollM, int unrollMToSubgroups,
-                   int unrollN, int unrollNToSubgroups,
-                   int unrollK) -> std::optional<DataTiledMMAAttr> {
-    if (!hasIntrinsic(target, intrinsic)) {
-      return std::nullopt;
+
+  //
+  // Step 1: select a MMAIntrinsic.
+  //
+  SmallVector<MMAIntrinsic> candidateIntrinsics = {
+      MMAIntrinsic::MFMA_F32_16x16x4_F32,
+      MMAIntrinsic::MFMA_F32_16x16x16_F16,
+      MMAIntrinsic::MFMA_I32_16x16x32_I8,
+  };
+  std::optional<MMAIntrinsic> intrinsic;
+  for (MMAIntrinsic candidateIntrinsic : candidateIntrinsics) {
+    if (!hasIntrinsic(target, candidateIntrinsic)) {
+      continue;
     }
-    auto candidate = DataTiledMMAAttr::get(
-        ctx, MMAIntrinsicAttr::get(ctx, intrinsic), /*unroll_m=*/unrollM,
-        /*unroll_m_to_subgroups=*/unrollMToSubgroups, /*unroll_n=*/unrollN,
-        /*unroll_n_to_subgroups=*/unrollNToSubgroups, /*unroll_k=*/unrollK);
-    auto [candidateLhs, candidateRhs, candidateOut] =
-        candidate.getABCElementTypes();
-    if (candidateLhs != lhs || candidateRhs != rhs || candidateOut != out) {
-      return std::nullopt;
+    auto [et0, et1, et2] =
+        MMAAttr::get(ctx, candidateIntrinsic).getABCElementTypes();
+    if (et0 != eTypes[0] || et1 != eTypes[1] || et2 != eTypes[2]) {
+      continue;
     }
-    return candidate;
-  };
-  if (auto m = match(MMAIntrinsic::MFMA_F32_16x16x4_F32, 8, 1, 2, 4, 4)) {
-    return m;
+    intrinsic = candidateIntrinsic;
+    break;
+  }
+  if (!intrinsic) {
+    return std::nullopt;
   }
-  if (auto m = match(MMAIntrinsic::MFMA_F32_16x16x16_F16, 8, 1, 2, 4, 2)) {
-    return m;
+
+  //
+  // Step 2: Select the unrolling factors for the generic case where there is no
+  //         narrow dimension.
+  //
+  // These hardcoded constants should become functions querying `target`.
+  //
+  // Target ISA preferred load instruction size, in bits.
+  const int kLoadInstructionBits = 128;
+  // Target ISA preferred number of subgroups per block to get full utilization.
+  const int kMaxSubgroups = 4;
```

Comment:
Why 'max' instead of 'min'? This is the minimum number of subgroups to saturate all SIMD units